src/utils/connect.js

```diff
@@ -0,0 +1,115 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+// @flow
+
+import * as React from 'react';
+import { connect } from 'react-redux';
+import type {} from 'react-redux';
+import type { Dispatch, State } from '../types/store';
+
+type MapStateToProps<OwnProps: Object, StateProps: Object> = (
+  state: State,
+  ownProps: OwnProps
+) => StateProps;
+
+type MapDispatchToProps<OwnProps: Object, DispatchProps: Object> =
+  | ((dispatch: Dispatch, ownProps: OwnProps) => DispatchProps)
+  | DispatchProps;
+
+type MergeProps<
+  StateProps,
+  DispatchProps: Object,
+  OwnProps: Object,
+  Props: Object
+> = (
+  stateProps: StateProps,
+  dispatchProps: DispatchProps,
+  ownProps: OwnProps
+) => Props;
+
+type ConnectOptions = {
+  pure?: boolean,
+  areStatesEqual?: boolean,
+  areOwnPropsEqual?: boolean,
+  areStatePropsEqual?: boolean,
+  areMergedPropsEqual?: boolean,
+  storeKey?: boolean,
+  withRef?: boolean,
+};
+
+export type ExplicitConnectOptions<
+  OwnProps: Object,
+  StateProps: Object,
+  DispatchProps: Object
+> = {
+  mapStateToProps?: MapStateToProps<OwnProps, StateProps>,
+  mapDispatchToProps?: MapDispatchToProps<OwnProps, DispatchProps>,
+  mergeProps?: MergeProps<
+    StateProps,
+    DispatchProps,
+    OwnProps,
+    {
+      ...OwnProps,
+      ...StateProps,
+      ...DispatchProps,
+    }
```

Comment:
do we want the exact object here?

Reply:
also, I believe this is not exactly right here: mergeProps will usually return another set of props.

I believe this set of Props (maybe we can call it MergeProps) could also be a generic parameter, with a default value of `{ ...OwnProps, ...StateProps, ...DispatchProps }` (or possibly the exact/readonly variation).

**that said** we don't use `mergeProps` anywhere at the moment, so maybe we can leave it for a future patch? wdyt?

Reply:
I made some changes to how this is implemented, so I'm not sure exactly how this plays in now. I'm fine deferring until it becomes an issue.